# Exercicis d'algoritmes de Classificació.
# Ejercicios de algoritmos de Classificación
# Classification Algorithm Exercises.

- Exercici 1

Crea almenys dos models de classificació diferents per intentar predir el millor les classes de l'arxiu adjunt.

- Ejercicio 1

Crea al menos dos modelos de clasificación diferentes para intentar predecir mejor las clases del archivo adjunto.

- Exercise 1

Create at least two different classification models to better predict the classes of the attached file.

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

1. Title of Database: Wine recognition data Updated Sept 21, 1998 by C.Blake : Added attribute information

2. Sources:
   (a) Forina, M. et al, PARVUS - An Extendible Package for Data       Exploration, Classification and Correlation. Institute of Pharmaceutical        and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy.

   (b) Stefan Aeberhard, email: stefan@coral.cs.jcu.edu.au
   (c) July 1991
3. Past Usage:

   (1)
   S. Aeberhard, D. Coomans and O. de Vel,
   Comparison of Classifiers in High Dimensional Settings,
   Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of
   Mathematics and Statistics, James Cook University of North Queensland.
   (Also submitted to Technometrics).

   The data was used with many others for comparing various
   classifiers. The classes are separable, though only RDA
   has achieved 100% correct classification.
   (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data))
   (All results using the leave-one-out technique)

   In a classification context, this is a well posed problem
   with "well behaved" class structures. A good data set
   for first testing of a new classifier, but not very
   challenging.

   (2)
   S. Aeberhard, D. Coomans and O. de Vel,
   "THE CLASSIFICATION PERFORMANCE OF RDA"
   Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of
   Mathematics and Statistics, James Cook University of North Queensland.
   (Also submitted to Journal of Chemometrics).

   Here, the data was used to illustrate the superior performance of
   the use of a new appreciation function with RDA.

4. Relevant Information:

   -- These data are the results of a chemical analysis of
      wines grown in the same region in Italy but derived from three
      different cultivars.
      The analysis determined the quantities of 13 constituents
      found in each of the three types of wines.

   -- I think that the initial data set had around 30 variables, but
      for some reason I only have the 13 dimensional version.
      I had a list of what the 30 or so variables were, but a.)
      I lost it, and b.), I would not know which 13 variables
      are included in the set.

   -- The attributes are (dontated by Riccardo Leardi, riclea@anchem.unige.it):

 	1) Alcohol

 	2) Malic acid

 	3) Ash

  4) Alcalinity of ash  

  5) Magnesium

  6) Total phenols

  7) Flavanoids

  8) Nonflavanoid phenols

  9) Proanthocyanins

  10)Color intensity

  11)Hue

  12)OD280/OD315 of diluted wines

  13)Proline            

5. Number of Instances:

 	class 1 59

  class 2 71

  class 3 48

6. Number of Attributes:

	13

7. For Each Attribute:

	All attributes are continuous

	No statistics available, but suggest to standardise
	variables for certain uses (e.g. for us with classifiers
	which are NOT scale invariant)

	NOTE: 1st attribute is class identifier (1-3)

8. Missing Attribute Values:

	None

9. Class Distribution: number of instances per class:


  class 1 59

  class 2 71

  class 3 48


In [3]:
df = pd.read_csv('/content/gdrive/MyDrive/DS/M7T01/wineData.txt', sep=',',header=None)

In [ ]:
df

0      1     2     3     4    5     6     7     8     9      10    11  \
0     1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29   5.64  1.04   
1     1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28   4.38  1.05   
2     1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81   5.68  1.03   
3     1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18   7.80  0.86   
4     1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82   4.32  1.04   
..   ..    ...   ...   ...   ...  ...   ...   ...   ...   ...    ...   ...   
173   3  13.71  5.65  2.45  20.5   95  1.68  0.61  0.52  1.06   7.70  0.64   
174   3  13.40  3.91  2.48  23.0  102  1.80  0.75  0.43  1.41   7.30  0.70   
175   3  13.27  4.28  2.26  20.0  120  1.59  0.69  0.43  1.35  10.20  0.59   
176   3  13.17  2.59  2.37  20.0  120  1.65  0.68  0.53  1.46   9.30  0.60   
177   3  14.13  4.10  2.74  24.5   96  2.05  0.76  0.56  1.35   9.20  0.61   

       12    13  
0    3.92  1065  
1    3.40  1050  
2    3.17  1185  
3    3.45  1480  
4    2.93   735  
..    ...   ...  
173  1.74   740  
174  1.56   750  
175  1.56   835  
176  1.62   840  
177  1.60   560  

[178 rows x 14 columns]

In [4]:
cols =[
    "class",
    "Alcohol",
    "Malic acid",
    "Ash",
    "Alcalinity of ash",
    "Magnesium",
    "Total phenols",
    "Flavanoids",
    "Nonflavanoid phenols",
    "Proanthocyanins",
    "Color intensity",
    "Hue",
    "OD280/OD315 of diluted wines",
    "Proline"
]


In [5]:
df.columns = cols
df.head(3)

class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0      1    14.23        1.71  2.43               15.6        127   
1      1    13.20        1.78  2.14               11.2        100   
2      1    13.16        2.36  2.67               18.6        101   

   Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier


In [7]:
# Dividir los datos en conjuntos de entrenamiento y prueba
X = df.drop("class", axis=1)
y = df["class"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [10]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


In [11]:
models = {
    'Logistic Regression': LogisticRegression(),
    'Random Forest': RandomForestClassifier(),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Support Vector Machine': SVC()

}

for model_name, model in models.items():
    model.fit(X_train, y_train)
    predictions = model.predict(X_test)

    accuracy = accuracy_score(y_test, predictions)
    confusion = confusion_matrix(y_test, predictions)
    report = classification_report(y_test, predictions)

    print(model_name + ":")
    print("Accuracy:", accuracy)
    print("Confusion Matrix:\n", confusion)
    print("Classification Report:\n", report)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Logistic Regression:
Accuracy: 0.9722222222222222
Confusion Matrix:
 [[13  1  0]
 [ 0 14  0]
 [ 0  0  8]]
Classification Report:
               precision    recall  f1-score   support

           1       1.00      0.93      0.96        14
           2       0.93      1.00      0.97        14
           3       1.00      1.00      1.00         8

    accuracy                           0.97        36
   macro avg       0.98      0.98      0.98        36
weighted avg       0.97      0.97      0.97        36

Random Forest:
Accuracy: 1.0
Confusion Matrix:
 [[14  0  0]
 [ 0 14  0]
 [ 0  0  8]]
Classification Report:
               precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00        14
           3       1.00      1.00      1.00         8

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36


- Exercici 2

Compara els models de classificació utilitzant la precisió (accuracy), una matriu de confusió i d’altres mètriques més avançades.

- Ejercicio 2

Compara los modelos de clasificación utilizando la precisión (accuracy), una matriz de confusión y otras métricas más avanzadas.

- Exercise 2

Compare the classification models using accuracy, a confusion matrix, and other advanced metrics.

En este informe, se muestra el rendimiento del modelo de Regresión Logística:
- La precisión del modelo es 0.972, lo que indica que el 97.2% de las predicciones son correctas.
- La matriz de confusión muestra cómo se clasificaron las instancias en las diferentes clases. Por ejemplo, el modelo clasificó correctamente 13 instancias de la clase 1, 14 instancias de la clase 2 y 8 instancias de la clase 3.
- El informe de clasificación proporciona información detallada sobre la precisión, recuperación y puntuación F1 para cada clase y en general. En general, el modelo muestra un rendimiento muy bueno en todas las métricas de evaluación.

Este modelo parece tener un rendimiento sólido en términos de precisión y recuperación para todas las clases.

Random Forest:
- La precisión del modelo es 1.0, lo que indica que el 100% de las predicciones son correctas.
- La matriz de confusión muestra que todas las instancias se clasificaron correctamente en sus respectivas clases.
- El informe de clasificación también muestra puntuaciones perfectas en precisión, recuperación y puntuación F1 para todas las clases y en general.

Este modelo parece tener un rendimiento excepcionalmente alto y preciso en la clasificación de las instancias en todas las clases.

K-Nearest Neighbors (KNN):
- La precisión del modelo es 0.7222222222222222, lo que indica que alrededor del 72.22% de las predicciones son correctas.
- La matriz de confusión muestra cómo se distribuyen las instancias clasificadas en las clases. Algunas instancias fueron clasificadas incorrectamente.
- El informe de clasificación proporciona más detalles sobre la precisión, la recuperación y la puntuación F1 para cada clase individual, así como sus valores promedio ponderados y macro. En este caso, el modelo tiene un rendimiento relativamente bueno para la clase 1, pero su rendimiento es menos satisfactorio para la clase 3.

Este modelo parece tener un rendimiento moderado en términos de precisión y puede beneficiarse de ajustes y optimizaciones.

Máquinas de Vectores de Soporte (SVM):
- La precisión del modelo es 0.8055555555555556, lo que indica que alrededor del 80.56% de las predicciones son correctas.
- La matriz de confusión muestra cómo se distribuyen las instancias clasificadas en las clases. El modelo tuvo algunos errores en la clasificación de las clases 2 y 3.
- El informe de clasificación proporciona más detalles sobre la precisión, la recuperación y la puntuación F1 para cada clase individual, así como sus valores promedio ponderados y macro. En este caso, el modelo tiene un rendimiento sólido para la clase 1, pero su rendimiento es menos consistente para las clases 2 y 3.

Este modelo podría beneficiarse de ajustes de hiperparámetros o técnicas de selección de características para mejorar su rendimiento.

- Exercici 3

Entrena’ls usant els diferents paràmetres que admeten per tal de millorar-ne la predicció.

- Ejercicio 3

Entrenalos utilizando los diferentes parámetros que admiten para intentar mejorar su predicción.

- Exercise 3

Train them using the different parameters they allow to improve their prediction.

In [16]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

In [ ]:
# Definir los modelos con sus hiperparámetros para GridSearchCV o RandomizedSearchCV
models = {
    'Logistic Regression': (LogisticRegression(), {
        'C': [0.1, 1, 10],
        'max_iter': [100, 200, 300]
    }),
    'Random Forest': (RandomForestClassifier(), {
        'n_estimators': [50, 100, 150],
        'max_depth': [None, 10, 20]
    }),
    'K-Nearest Neighbors': (KNeighborsClassifier(), {
        'n_neighbors': [3, 5, 7],
        'weights': ['uniform', 'distance']
    }),
    'Support Vector Machine': (SVC(), {
        'C': [0.1, 1, 10],
        'kernel': ['linear', 'rbf']
    })
}

# Realizar GridSearchCV o RandomizedSearchCV para cada modelo
for model_name, (model, param_grid) in models.items():
    print("Training", model_name)

    # Realizar búsqueda de parámetros
    search = GridSearchCV(model, param_grid, cv=5)  # Cambia a RandomizedSearchCV si lo prefieres

    # Entrenar el mejor modelo encontrado
    search.fit(X_train, y_train)
    best_model = search.best_estimator_

    # Predecir en el conjunto de prueba
    predictions = best_model.predict(X_test)

    # Evaluar el modelo
    accuracy = accuracy_score(y_test, predictions)
    confusion = confusion_matrix(y_test, predictions)
    report = classification_report(y_test, predictions)

    # Imprimir resultados
    print(model_name + ":")
    print("Best Parameters:", search.best_params_)
    print("Accuracy:", accuracy)
    print("Confusion Matrix:\n", confusion)
    print("Classification Report:\n", report)
    print("------------")

Training Logistic Regression


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Logistic Regression:
Best Parameters: {'C': 0.1, 'max_iter': 300}
Accuracy: 1.0
Confusion Matrix:
 [[14  0  0]
 [ 0 14  0]
 [ 0  0  8]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00         8

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

------------
Training Random Forest
Random Forest:
Best Parameters: {'max_depth': 10, 'n_estimators': 150}
Accuracy: 1.0
Confusion Matrix:
 [[14  0  0]
 [ 0 14  0]
 [ 0  0  8]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00         8

    accuracy                           1

Los resultados de la optimización de parámetros para cada modelo son los siguientes:

Logistic Regression:

    Mejores parámetros: {'C': 0.1, 'max_iter': 300}
    Accuracy: 1.0
    
Random Forest:

    Mejores parámetros: {'max_depth': 10, 'n_estimators': 150}
    Accuracy: 1.0
    
K-Nearest Neighbors:

    Mejores parámetros: {'n_neighbors': 7, 'weights': 'distance'}
    Accuracy: 0.75
    
Support Vector Machine:

    Mejores parámetros: {'C': 0.1, 'kernel': 'linear'}
    Accuracy: 1.0
    

- Exercici 4

Compara el seu rendiment fent servir l’aproximació traint/test o cross-validation.

- Ejercicio 4

Compara su rendimiento utilizando la aproximación entrenamiento/prueba o validación cruzada.

- Exercise 4

Compare their performance using the train/test approach or cross-validation.

In [19]:
# Definir los modelos con sus hiperparámetros para GridSearchCV o RandomizedSearchCV
models = {
    'Logistic Regression': LogisticRegression(C=0.1, max_iter=300),
    'Random Forest': RandomForestClassifier(max_depth=20, n_estimators=100),
    'K-Nearest Neighbors': KNeighborsClassifier(n_neighbors=7, weights='distance'),
    'Support Vector Machine': SVC(C=0.1, kernel='linear')
}

# Realizar validación cruzada para cada modelo
for model_name, model in models.items():
    print("Training", model_name)

    # Realizar validación cruzada con 5 folds
    cv_scores = cross_val_score(model, X_train, y_train, cv=5)

    # Imprimir resultados
    print(model_name + ":")
    print("Promedio de Validación Cruzada:", np.mean(cv_scores))
    print("------------")

Training Logistic Regression


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Logistic Regression:
Promedio de Validación Cruzada: 0.9504926108374384
------------
Training Random Forest
Random Forest:
Promedio de Validación Cruzada: 0.9716748768472907
------------
Training K-Nearest Neighbors
K-Nearest Neighbors:
Promedio de Validación Cruzada: 0.7187192118226602
------------
Training Support Vector Machine
Support Vector Machine:
Promedio de Validación Cruzada: 0.9576354679802955
------------


Logistic Regression muestra un accuracy perfecta en el conjunto de entrenamiento/prueba, lo que sugiere que está ajustando muy bien los datos de entrenamiento. Sin embargo, en la validación cruzada, donde se evalúa en conjuntos de datos diferentes, la precisión promedio es ligeramente más baja (0.9505). Esto indica que el modelo puede tener un rendimiento un poco más bajo en datos no vistos, pero aún es muy sólido.

Random Forest también tiene un accuracy perfecta en el conjunto de entrenamiento/prueba. Sin embargo, en la validación cruzada, su rendimiento promedio es incluso mejor (0.9717), lo que sugiere que es un modelo altamente generalizable y estable en datos no vistos.

K-Nearest Neighbors tiene un accuracy de 0.75, lo que indica un rendimiento moderado. Sin embargo, en la validación cruzada, su rendimiento promedio es más bajo (0.7187), lo que sugiere que el modelo puede no ser tan confiable en datos no vistos y puede estar sobreajustando en ciertos conjuntos de entrenamiento.

SVM muestra un accuracy perfecta en el conjunto de entrenamiento/prueba. Sin embargo, en la validación cruzada, su rendimiento promedio es ligeramente más bajo (0.9576), lo que sugiere que puede haber una pequeña pérdida de rendimiento en datos no vistos, pero aún es un modelo confiable.

- Exercici 5

Aplica algun procés d'enginyeria per millorar els resultats (normalització, estandardització, mostreig...).

- Ejercicio 5

Aplica algún proceso de ingeniería para mejorar los resultados (normalización, estandarización, muestreo...).

- Exercise 5

Apply some engineering processes to enhance the results (normalization, standardization, sampling...).


In [20]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler
from imblearn.combine import SMOTEENN
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.pipeline import make_pipeline

# Dividir el conjunto de datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Definir los modelos con sus hiperparámetros
models = {
    'Logistic Regression': LogisticRegression(),
    'Random Forest': RandomForestClassifier(),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Support Vector Machine': SVC()
}

# Aplicar preprocesamiento y entrenar los modelos
for model_name, model in models.items():
    print("Training", model_name)

    # Crear un pipeline con preprocesamiento y modelo
    pipeline = make_pipeline(StandardScaler(), model)

    # Realizar validación cruzada
    cv_scores = cross_val_score(pipeline, X_train, y_train, cv=5)

    # Entrenar el modelo final
    pipeline.fit(X_train, y_train)
    predictions = pipeline.predict(X_test)

    # Evaluar el modelo
    accuracy = accuracy_score(y_test, predictions)

    # Imprimir resultados
    print(model_name + ":")
    print("Promedio de Validación Cruzada:", cv_scores.mean())
    print("Accuracy:", accuracy)
    print("------------")


Training Logistic Regression
Logistic Regression:
Promedio de Validación Cruzada: 0.993103448275862
Accuracy: 0.9722222222222222
------------
Training Random Forest
Random Forest:
Promedio de Validación Cruzada: 0.9862068965517242
Accuracy: 1.0
------------
Training K-Nearest Neighbors
K-Nearest Neighbors:
Promedio de Validación Cruzada: 0.9650246305418719
Accuracy: 0.9722222222222222
------------
Training Support Vector Machine
Support Vector Machine:
Promedio de Validación Cruzada: 0.9862068965517242
Accuracy: 0.9722222222222222
------------


Logistic Regression:

    El promedio de validación cruzada es aproximadamente 0.993, lo que sugiere que el modelo es capaz de generalizar bien en diferentes conjuntos de datos de entrenamiento.
    La accuracy en el conjunto de prueba es del 0.972, lo que significa que el modelo predice correctamente la clase en un 97.2% de las instancias del conjunto de prueba.

Random Forest:

    El promedio de validación cruzada es aproximadamente 0.986, indicando que el modelo tiene una buena capacidad de generalización.
    La accuracy en el conjunto de prueba es del 1.0, lo que significa que el modelo predice todas las instancias correctamente en el conjunto de prueba.

K-Nearest Neighbors:

    El promedio de validación cruzada es aproximadamente 0.965, lo que sugiere una capacidad razonable de generalización.
    La accuracy en el conjunto de prueba es del 0.972, lo que indica que el modelo tiene un rendimiento similar al conjunto de entrenamiento.

Support Vector Machine:

    El promedio de validación cruzada es aproximadamente 0.986, lo que indica una buena capacidad de generalización.
    La acccuracy en el conjunto de prueba es del 0.972, lo que demuestra que el modelo tiene un buen rendimiento en el conjunto de prueba.